src/scheduler/pbs_sched.c

```diff
@@ -1273,3 +1188,140 @@ main(int argc, char *argv[])
 	log_close(1);
 	exit(0);
 }
+
+/**
+ *
+ * @brief
+ *		Compares two sched commands(command in cache vs new command coming from socket) and finds out
+ *	        whether they are identical or not. Also replaces command in cache with the new command if new command
+ *	        is superset.
+ *	        Since this is a separate function more filtering criteria for Sched commands can be easily added. For example
+ *              if a new sched command is introduced or priority of an existing command has changed this function has to be
+ *              revisited.
+ *
+ * @param[in]	cmd_cache	-	command in Sched's cache
+ * @param[in]	new_cmd		-	command from socket buffer
+ *
+ *
+ * @return	int
+ * @retval	1	-	if both command are duplicate
+ * @reval	0	-	if not duplicate
+ *
+ */
+static int
+compare_modify_sched_cmd(sched_cmd *cmd_cache, sched_cmd *new_cmd)
+{
+	if (new_cmd->cmd == SCH_SCHEDULE_RESTART_CYCLE )
+		return 1;
+	
+	if (new_cmd->cmd == SCH_SCHEDULE_AJOB &&
+		(cmd_cache->cmd != SCH_SCHEDULE_AJOB && cmd_cache->cmd != SCH_CONFIGURE &&
+		 cmd_cache->cmd != SCH_RULESET && cmd_cache->cmd != SCH_ERROR)) {
+		return 1;
+	} else if (cmd_cache->cmd == SCH_SCHEDULE_AJOB &&
+		(new_cmd->cmd != SCH_SCHEDULE_AJOB && new_cmd->cmd != SCH_CONFIGURE &&
+		 new_cmd->cmd != SCH_RULESET && new_cmd->cmd != SCH_ERROR)) {
+		cmd_cache->cmd = new_cmd->cmd;
+		cmd_cache->jid = NULL;	
+		return 1;	
+	}
+
+	if (cmd_cache && new_cmd) {
+		if (cmd_cache->jid == NULL && new_cmd->jid == NULL) {
+			if (cmd_cache->cmd == new_cmd->cmd)
+				return 1;
+		} else {
+			if (cmd_cache->cmd == new_cmd->cmd && !strcmp(cmd_cache->jid, new_cmd->jid))
+				return 1;
+		}
+	}
+	return 0;
+}
+
+/**
+ * @brief Filters scheduler commands queue according to the criteria given in the function
```

Comment:
I think that since in our case commands have always been internal to PBS, we can renumber them in the priority order and use their id as an index in commands array.
Like I'll probably make RESTART_CYCLE as number 0, SCH_SCHEDULE_AJOB as 1.... and SCH_SCHEDULE_NEW as number 17.
So the highest priority jobs will be the first in the array and others will be behind it.
When you set an index based on what you receive, you can also turn off other similar commands like SCH_SCHEDULE_STARTQ will unset SCH_SCHEDULE_NEW etc.
Scheduler can then go through the array of commands and start to honor them one after another.
The only exception I think is in the case of SCH_SCHEDULE_AJOB. In that case, you will have to append the job id into another list and not turn off the bit in commands array until all are done.

Reply:
As we know as of today Server marks scheduling commands as super high priority, medium priority and normal priority and sends them accordingly. Current PR does not change this behaviour. 

The job of this function filter_sched_cmds is 
- Filter duplicate scheduling commands. Duplicate commands are possible since multiple servers can kick off scheduling cycle. For example if we have two servers and both of them issuing a SCH_SCHEDULE_NEW. Scheduler needs to discard one command.
- Similarly if a scheduler has SCH_SCHEDULE_NEW and SCH_SCHEDULE_AJOB at a time it discards  SCH_SCHEDULE_AJOB
- If two servers sends SCH_SCHEDULE_AJOB each and we don't have any commands in scheduler's queue both of these commands are retained because jobids are different.
- All of this criteria is hidden inside the function compare_modify_sched_cmd.

I have added detailed descriptions in the code itself in these functions. Please take a look.

Reply:
You can't discard SCHEDULE_AJOB.  When the scheduler receives a SCHEDULE_AJOB cycle, it does a special cycle for a specific job (and only that job).  This job gets super high priority during this cycle.  This includes doing things like ignoring limits and doing preemption with the highest possible preemption priority.  *EVERY* qrun job must have its own cycle.  We can't bunch these together or they lose their special priority.

In your previous testing, you just made sure that the jobs ran.  You didn't check to make sure they got the priority they needed and we overrode the policies they required.  Also, a qrun job has a special IFL call to return an error message if that job fails to run.  If you bunch these up, you will lose these (look at pbs_defschreply() at the end of scheduling_cycle())

Reply:
I feel the current PR is already changing the behavior in some way. The scheduler is now taking a decision that out of a handful of requests that has received from different servers which is the one it needs to honor and which is the one it can avoid. Isn't this prioritizing the commands?

In relation to the second point you mentioned, I think that is wrong. We have to preserve the job id because these jobs are treated with elevated preemption priority.
The way you have it now, if a qrun request gets overwritten by a normal sched cycle request, the job that is being qrun will not be treated in special way and may not preempt.

Reply:
So to keep things "the same" between a single server, and multi-server, we can do some command filtering.  Right now, the server keeps two command variables, and a list of all qrun requests.  The two command variables are a high priority one and a low priority one.  The high priority one is for things like SCH_CONFIGURE.  The normal priority one is for running scheduling cycles.  If multiple normal priority commands come in, the current one overwrites the previous one.  For the most part, all the normal scheduling commands will run a cycle.  If we wanted to extend this to multi-server, we could filter all the normal priority scheduling commands down to just one.  Until the server will send all non-duplicate commands, the scheduler can't trust the command anyway.  If it receives a SCH_SCHEDULE_NEW, it can't trust that there wasn't a SCH_SCHEDULE_TERM that came in right behind it.  If we want to start sending all non-duplicate commands from all servers, then the scheduler can make inferences based on the command.  Until we decide to do that, they all just get dumped into a 'run a scheduling cycle' bucket.

Reply:
> You can't discard SCHEDULE_AJOB. When the scheduler receives a SCHEDULE_AJOB cycle, it does a special cycle for a specific job (and only that job). This job gets super high priority during this cycle. This includes doing things like ignoring limits and doing preemption with the highest possible preemption priority. _EVERY_ qrun job must have its own cycle. We can't bunch these together or they lose their special priority.
> 
> In your previous testing, you just made sure that the jobs ran. You didn't check to make sure they got the priority they needed and we overrode the policies they required. Also, a qrun job has a special IFL call to return an error message if that job fails to run. If you bunch these up, you will lose these (look at pbs_defschreply() at the end of scheduling_cycle())

Thanks @bhroam for the detailed explanation. Modified the code changes so as not to discard or replace SCHEDULE_AJOB. Please look into my latest commit.

Reply:
> So to keep things "the same" between a single server, and multi-server, we can do some command filtering. Right now, the server keeps two command variables, and a list of all qrun requests. The two command variables are a high priority one and a low priority one. The high priority one is for things like SCH_CONFIGURE. The normal priority one is for running scheduling cycles. If multiple normal priority commands come in, the current one overwrites the previous one.
> 
> For the most part, all the normal scheduling commands will run a cycle. If we wanted to extend this to multi-server, we could filter all the normal priority scheduling commands down to just one. Until the server will send all non-duplicate commands, the scheduler can't trust the command anyway. If it receives a SCH_SCHEDULE_NEW, it can't trust that there wasn't a SCH_SCHEDULE_TERM that came in right behind it. If we want to start sending all non-duplicate commands from all servers, then the scheduler can make inferences based on the command. Until we decide to do that, they all just get dumped into a 'run a scheduling cycle' bucket.

Thank you @bhroam. It looks like you are suggesting to divide around 18 scheduling commands that we have right now into 3 buckets/groups 

Bucket-1: SCH_CONFIGURE/SCH_QUIT
Bucket-2: List of qrun requests
Bucket-3: Rest all commands to say "Run a sched cycle"

So Server will only send one command for each group. At the most Scheduler gets three types of commands and it filters the duplicates when there are multiple servers.  

I think these changes are really good to do which can actually go independently after the current PR. Hope this fine. Following are the reasons that made me to think in this direction.

- In the current PR I am not changing the number of scheduling commands and the way server sends commands to scheduler    i.e. I am not having any changes in Server whereas the proposed changes have lot of code changes in both server and scheduler side.
- The code changes done in this PR like making Scheduler ready for Multi-Server,  filtering sched commands mostly won't change even after the proposed changes happen after the current PR. Also completing the current PR as a first step helps other Multi-Server PRs which are dependent on this like IFL aggregation, implementing a scheduling policy that gives first preference to the local nodes, "Run and Movejob to other servers changes" in case if the server does not have the nodes associated with a job.